### ISSUE TITLE
Run Command for Offline Users

### ISSUE THREAD

**Issue by nxfifteen** [WooMinecraft/WooMinecraft#132]
Would it possible to run commands when users are offline? Forexample the 'pardon' command

I'm not an Java programmer, but my inital thought was a `system` username that Minecraft plug would recognise and run the command without a specific user being online

**Comment by JayWood** [comment]
@nxfifteen this is possible if they've been on your server at least once. However, at this time, we force the user to be online in order to run the command.

At this time I believe there's either `run while online` or `run while offline` and currently in the plugin there's no flags to say "run this even if he/she is offline" so I'll look into that for you.

**Comment by rbtkooijman** [comment]
I have another example why this would be preferable. Not being able to execute commands for offline players makes it impossible to use WooMinecraft with a permission based whitelist tool.

I am currently using PermissionsEX with the Modifyworld extension. Modifyworld has the 'modifyworld.login' node, which grants a player access to the server if given to them. Even though PermissionEX can add usernames or uuids to its permissions file even when the player is offline/hasn't been online before (thanks to Vault), this is not currently possible because WooMinecraft checks the player's availability and prevents the command from executing if the player is offline, correct?

**Comment by JayWood** [comment]
@rbtkooijman you're correct, this is not currently possible as the code uses `getPlayerExact()` - https://github.com/WooMinecraft/WooMinecraft/blob/master/src/main/java/com/plugish/woominecraft/WooMinecraft.java#L161

I do plan to allow for offline commands, but as of late my focus has been on fully releasing the multi-server support WP Plugin Update

So, this is on my radar.

**Comment by dordsor21** [comment]
`if(player.hasPlayedBefore()`

**Comment by nxfifteen** [comment]
@dordsor21 Sadly nothing is as easy as it should be. I just did a quick check running some command from my server terminal without being logged on and there are a limited subset of command you can run. For example, you can 'ban' and 'parton' but you cant 'give', 'kick' or 'xp' - it just throws the error
`Player 'xxxxx' cannot be found`

My first thoughts are ether there has to be a dictonary of command that can be run while offline, or there would need to be some error handling to detect when the command didnt work. Both of which have problems.

The first option, would probably have to be somewhere in the config file because if it was hardcoded anyone wanting to support a command for a specific plugin would have to rebuild the jar file.

The problem I see with the second option is the same 'give' command could be run every twenty minutes for days or weeks if the user isnt online all that time which for big servers could create unessesary lag/load

**Comment by JayWood** [comment]
@dordsor21 as @nxfifteen states, there are only certain commands that can be ran while a user is offline. Blanket-accepting commands while a user is offline can cause problems.

This enhancement has to happen in two places, on the WordPress side, and on the Minecraft side.

My thoughts for this would be to provide a checkbox on each command row when adding commands on the WordPress side and label it something like `Offline Command` with a tool tip explaining to the user what known commands are allowed to be ran offline.

This would require a modification of the JSON feed provided from the server, it currently looks something like this:
```
{
    "data": {
        "Phyrax": {
            "159": [
                "Test Command 2",
                "Just one variation"
            ]
        }
    },
    "success": true
}
```

Simply put, for each user's orders, I could separate the 'online' vs 'offline' command sets like so:

```
{
    "data": {
        "Phyrax": {
        	"online": {
        	     "159": [
	                "An offline command"
            	     ]
        	},
        	"offline": {
        	     "159": [
        		"An online command"
        	     ]
        	}
        }
    },
    "success": true
}
```

Once that's complete, I'll need to build two command executors for each, sorta. Basically use one method for offline users, and one for online ( to verify they're online )

**Comment by nxfifteen** [comment]
@JayWood that definitly seems like a better idea than mine. Putting the deciding of what command are offline and what are online on the Wordpress side would definitly make it more powerfull for user to customise. 

Even in my first idea of having in the plugin config folder would more than likely have required a full server restart each time a change was made

**Comment by JayWood** [comment]
@nxfifteen like I said before, it's on my radar 😄 just need to find time to sit down and focus on it haha, been thinking about this and other enhancements for weeks.

**Comment by nxfifteen** [comment]
I'm not great with Java, but I'm a confident PHP/Wordpress coder so if you are happy sharing some of your ideas as issues I'm more than happy to fork the Wordpress plugin and submit some patches for your approval

**Comment by spannerman79** [comment]
@JayWood Enjin plugin uses [TwoTuxLib](https://dev.bukkit.org/projects/tuxtwolib) (javdoc http://javadocs.yu8.me/tuxtwolib/ ) to handle offline players - maybe you could look at doing the same? I know depending on a third party plugin/library might not be ideal but that "wheel" has already been invented.

Even more so for people that use Enjin and want a better alternative

**Comment by JayWood** [comment]
@spannerman79 It's worth taking a look, and I'm already using a 3rd party file for HTTP requests simplification.

**Comment by dordsor21** [comment]
You could make the command  `online:give player <item>` and `offline:perms player set` with a `split(":")` as an easier fix? Not sure how easy implementing new stuff like that into WP is. And yeah, I forgot about commands like that, was only thinking of permissions as that's what I do with donations, my bad!

**Comment by cdmvirt** [comment]
This would be great as we use you plugin to unban users but if they are banned and cant get in then no way to execute the "unban %s" command.

**Comment by spannerman79** [comment]
It has been awhile since I last looked at this, there is a plugin that can handle and call the correct UUID if the player is online or offline - assuming that player has been on the server at least once within the cache expiry time

https://github.com/KaiKikuchi/UUIDProvider

Requires MySQL but that shouldn't be an issue for wordpress users

**Comment by dordsor21** [comment]
so it's a worse way of using Bukkit.getOfflinePlayer(p).getUniqueID()

**Comment by spannerman79** [comment]
Not within a bungee network and you have more then just one server accessing the mojang api.  if you want skin blobs (which is needed for bungee) you are heavily rate limited - thus needing a cache system running.

**Comment by dordsor21** [comment]
right, this had nothing to do with bungee lmao. Could always use the mojang servers

**Comment by spannerman79** [comment]
> right, this had nothing to do with bungee lmao

https://github.com/WooMinecraft/WooMinecraft/issues/36#issuecomment-172866488

Context matters

**Comment by dordsor21** [comment]
...k

**Comment by marcopaggioro** [comment]
Hi, any news about that?

**Comment by JayWood** [comment]
Will be done with WooMinecraft/woominecraft-wp#50
